crates/core/src/sql/compiler.rs

```diff
@@ -1075,4 +1078,49 @@ mod tests {
         assert!(compile_sql(&db, &db.begin_tx(), "select * from B join A on B.y = A.x").is_ok());
         Ok(())
     }
+
+    #[test]
+    fn compile_type_check() -> ResultTest<()> {
+        let db = TestDB::durable()?;
+        db.create_table_for_test(
+            "PlayerState",
+            &[("entity_id", AlgebraicType::U64)],
+            &[(0.into(), "entity_id")],
+        )?;
+        db.create_table_for_test(
+            "EnemyState",
+            &[("entity_id", AlgebraicType::I8)],
+            &[(0.into(), "entity_id")],
+        )?;
+        db.create_table_for_test(
+            "FriendState",
+            &[("entity_id", AlgebraicType::U64)],
+            &[(0.into(), "entity_id")],
+        )?;
+        let sql = "SELECT * FROM PlayerState WHERE entity_id = '161853'";
+
+        // Should fail with type mismatch for selections and joins.
+        //
+        // TODO: Type check other operations deferred for the new query engine.
+
+        assert_eq!(
+            compile_sql(&db, &db.begin_tx(), sql).map_err(|e| e.to_string()),
+            Err("SqlError: Type Mismatch: `table#4097.col#0: Some(Builtin(U64))` != `\"161853\": Some(Builtin(String))`, executing: `SELECT * FROM PlayerState WHERE entity_id = '161853'`".into())
```

Comment:
Do you want to assert on the error message, or would the error type suffice?

Reply:
I mean to see how the error will look.